README.md
[English](https://github.com/aromajoin/controller-sdk-java) / [日本語](README-JP.md)

# Controller SDK for Java &middot; [ ![Download](https://api.bintray.com/packages/aromajoin/maven/com.aromajoin.sdk%3Ajvm/images/download.svg) ](https://bintray.com/aromajoin/maven/com.aromajoin.sdk%3Ajvm/_latestVersion) [![License](https://img.shields.io/badge/license-Apache%202-4EB1BA.svg?style=flat-square)](https://www.apache.org/licenses/LICENSE-2.0.html)

**A pure Java library version of AromaShooter Controller SDK which is used to communicate with [Aroma Shooter devices](https://aromajoin.com/hardware/shooters/aroma-shooter-1)**  

# Table of Contents
1. [Supported devices](#supported-devices)  
2. [Prerequisites](#prerequisites)
3. [Installation](#installation)
4. [Usage](#usage)
    * [Setup](#setup)
    * [Scan and connect device](#scan-and-connect-device)
    * [Connected devices](#connected-devices)
    * [Diffuse](#diffuse)
    * [Stop](#stop)
    * [Disconnect](#disconnect)
5. [Issues](#issues)
7. [License](#license)


## Supported devices
* Aroma Shooter 1 USB
* Aroma Shooter 1 RS-485


## Prerequisites
* JRE version: >= 1.8+
* Donwload and install [drivers](http://www.ftdichip.com/FTDrivers.htm) based on your OS.

## Installation
### Gradle

1. Firstly, add our repository on the top of your [rootProject]/build.gradle.
```gradle
repositories {
    // ... other repositories
    maven { url "https://dl.bintray.com/aromajoin/maven/" }
}
```
2. Then, add `controller-sdk` dependency.
```gradle
dependencies {
    // ... other dependencies
    compile 'com.aromajoin.sdk:jvm:2.x.x'
}
```
### Maven
```xml
<dependency>
  <groupId>com.aromajoin.sdk</groupId>
  <artifactId>jvm</artifactId>
  <version>2.x.x</version>
  <type>pom</type>
</dependency>
```
### Binary files (.jar)
1. Directly download the latest *.jar file from [Bintray.](https://bintray.com/aromajoin/maven/com.aromajoin.sdk%3Ajvm#files/com/aromajoin/sdk/jvm) 
2. Add it into your project's build path.

## Usage
For details, please check our [Sample project](https://github.com/aromajoin/controller-sdk-java/tree/master/Sample) (Recommended way.)

### Setup
```java
// Initialize an USB Aroma Shooter instance
USBASController usbController = new USBASController();
```
### Scan and connect device

```java
usbController.scanAndConnect(new DiscoverCallback() {
    @Override
    public void onDiscovered(List<AromaShooter> aromaShooters) {
        for(AromaShooter aromaShooter : aromaShooters){
            // Detected Aroma Shooter.
        }
    }

    @Override
    public void onFailed(String msg) {
       // Failed on scanning.
    }
});
```

### Connected devices
```java
List<AromaShooter> connectedDevices = usbController.getConnectedDevices();
```

### Diffuse 
```java
/**
 * Diffuses aroma at device's ports.
 * @param duration     diffusing duration in milliseconds.
 * @param booster      whether booster is used or not.
 * @param ports        cartridge numbers to diffuse aroma. Value: 1 ~ 6.
 */
// For example, the following codes will diffuse aroma at cartridge 2 and 5 fro 3 seconds.
usbController.diffuseAll(3000, true, 2, 5);
```
### Stop
Stop all ports of current connected devices if they are diffusing 
```java
usbController.stopAllPorts();
```
### Disconnect
Disconnect all devices.
```java
usbController.disconnectAll();
```
## Issues
**If you get any issues or require any new features, please create a [new issue](https://github.com/aromajoin/controller-sdk-java/issues).**


## License
    The Apache License (Apache)
    Copyright (c) 2017 Aromajoin Corporation

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
    SOFTWARE.


CHANGELOG.md
Change Log
==========

Version 2.1.0 (2018-07-30)
--------------------------
* Add intensity control function
* Limit the diffuse duration to 10 sec

Version 2.0.9 *(2017-12-20)*
----------------------------
* Improve annoying PortInUseException handling part

Version 2.0.8 *(2017-12-15)*
----------------------------
* Downgrade nsjavaserial library from version 3.12.1 to 3.11.0
(Due to error on Raspberry Pi reported here: https://github.com/NeuronRobotics/nrjavaserial/issues/84)

Version 2.0.7 *(2017-12-08)*
----------------------------
* Refactor com.aromajoin.sdk.jvm.usb connection part codes.

Version 2.0.0 *(2017-08-18)*
----------------------------
* First release


Sample/src/main/java/Sample.java

/*
 * Copyright (c) 2017 Aromajoin Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations under the License.
 */

import com.aromajoin.sdk.core.callback.ConnectCallback;
import com.aromajoin.sdk.core.device.AromaShooter;
import com.aromajoin.sdk.jvm.rs485.RS485ASController;
import com.aromajoin.sdk.jvm.rs485.model.RS485AromaShooter;
import com.aromajoin.sdk.jvm.usb.DiscoverCallback;
import com.aromajoin.sdk.jvm.usb.USBASController;
import com.aromajoin.sdk.jvm.usb.model.USBAromaShooter;
import gnu.io.NRSerialPort;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public class Sample {
    private static Sample ourInstance = new Sample();

    public static Sample getInstance() {
        return ourInstance;
    }

    private USBASController usbASController;
    private RS485ASController rs485ASController;
    private List<USBAromaShooter> connectedUSBAromaShooters;
    private List<RS485ASController> connectedRS485AromaShooters;

    private Sample() {
        usbASController = new USBASController();
        rs485ASController = new RS485ASController();
        connectedUSBAromaShooters = new ArrayList<>();
        connectedRS485AromaShooters = new ArrayList<>();
    }

    public USBASController getUsbASController(){
        return this.usbASController;
    }

    public RS485ASController getRs485ASController() {
        return rs485ASController;
    }

    public void runTest(){
        scanUSBAromaShooters();

        USBAromaShooter aromaShooter = new USBAromaShooter("/dev/ttyUSB0");
        aromaShooter.connect();

        String serial = "ASN1RA0001";
        String portName = "/dev/ttyUSB0";
        RS485AromaShooter rs485AromaShooter = new RS485AromaShooter(serial, portName);
        rs485AromaShooter.connect();
    }

    public static void main(String args[]) {
        System.out.println("======= Java Sample of Controller SDK =========");
        Sample sample = Sample.getInstance();
        sample.runTest();
        System.out.println("======= DONE =========");
    }

    private void scanUSBAromaShooters() {
        // Scanning USB Aroma Shooters
        this.getUsbASController().scan(new DiscoverCallback() {
            @Override
            public void onDiscovered(List<USBAromaShooter> list) {
                List<USBAromaShooter> aromaShooters =  list;
                if(aromaShooters.isEmpty()){
                    System.out.println("There is no USB Aroma Shooters!");
                    return;
                }
                for(USBAromaShooter aromaShooter : aromaShooters) {
                    connectUSBAromaShooter(aromaShooter);
                }
            }
            @Override
            public void onFailed(String s) {
                System.out.println(s);
            }
        });
    }

    private void connectUSBAromaShooter(USBAromaShooter usbAromaShooter){
        if(usbAromaShooter.isAvailable()){
            usbAromaShooter.connect();
        }
    }

    private void diffuseUSBAromaShooter(USBAromaShooter usbAromaShooter){
        if(usbAromaShooter.isAvailable() && usbAromaShooter.isUSBAromaShooter()){
            int[] ports = new int[2];
            ports[0] = 1;
            ports[1] = 3;
            usbAromaShooter.diffuse(1000, true, ports);
        }
    }

    private void diffuseRS485AromaShooter(RS485AromaShooter rs485AromaShooter){
        if(rs485AromaShooter.isAvailable() && rs485AromaShooter.isRS485AromaShooter()){
            int[] ports = new int[2];
            ports[0] = 1;
            ports[1] = 3;
            rs485AromaShooter.diffuse(1000, true, ports);
        }
    }
}


Sample/src/main/java/USBSample.java

/*
 * Copyright (c) 2017 Aromajoin Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations under the License.
 */

import com.aromajoin.sdk.core.callback.DisconnectCallback;
import com.aromajoin.sdk.core.device.AromaShooter;
import com.aromajoin.sdk.jvm.usb.USBASController;

import java.util.List;

public class USBSample {

    public void runTest(){
        USBASController usbController = new USBASController("/dev/ttyUSB0");
        usbController.diffuseAll(5000, true, 1,3,5);
        usbController.disconnectAll();
    }

    public static void main(String args[]) {
        System.out.println("======= Java USBSample of Controller SDK =========");
        USBSample sample = new USBSample();
        sample.runTest();
    }
}


Sample/src/main/java/AromaShooterUSBSample.java
import com.aromajoin.sdk.core.device.AromaShooter;
import com.aromajoin.sdk.jvm.DiscoverCallback;
import com.aromajoin.sdk.jvm.usb.USBASController;
import java.util.List;

public class AromaShooterUSBSample {
  public static void main(String args[]) {
    System.out.println("======= A sample of Controller SDK for Aroma Shooter USB=========");
    //testWithSpecificPort();
    testWithAutomaticScan();
    System.out.println("======= DONE =========");
  }

  /**
   * You can define AromaShooter USB connected port.
   */
  private static void testWithSpecificPort() {
    USBASController usbController = new USBASController("/dev/tty.usbserial-AH03I8XI");
    usbController.diffuseAll(3000, true, 2, 3);
    usbController.disconnectAll();
  }

  /**
   * For more convenient, you can scan all connected Aroma Shooter.
   */
  private static void testWithAutomaticScan() {
    USBASController usbController = new USBASController();
    usbController.scanAndConnect(new DiscoverCallback() {
      @Override public void onDiscovered(List<AromaShooter> aromaShooters) {
        for (AromaShooter aromaShooter : aromaShooters) {
          System.out.println("AromaShooter: " + aromaShooter.getSerial());
        }
        usbController.diffuseAll(3000, true, 2, 3);
      }

      @Override public void onFailed(String msg) {

      }
    });
  }
}


AromaShooterControllerSample/src/aromashootercontrollertest/Main.java
/**
 * 
 */
package aromashootercontrollertest;

import com.aromajoin.www.aromashooter.*;

/**
 * @author hanh
 *
 */
public class Main {

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		testUSBDevice();
	}
	
	private static void testUSBDevice(){
		// アロマシューターが接続中のシリアルポートを指定します（例：COM3）．
        String portName = "/dev/ttyUSB0";

        // 指定したシリアルポートに接続中の機器がアロマシュータ－か否かを確認します．
        if (AromaShooter.isAromaShooter("", portName)) {
            System.out.println(portName + " に接続中の機器はアロマシュータです．");
        } else {
            System.out.println(portName + " に接続中の機器はアロマシュータではありません．");
            return;
        }

        // アロマシュータークラスのインスタンスを生成します
        AromaShooter as = new AromaShooter(portName);
        System.out.println(as.toString());

        // 香りを噴射します
        int durationMilliSec = 3000;// 噴射期間(ミリ秒)
        double density = 0.6;// 香りの濃度(0～1.0)
        // 噴射速度は、BLOWING_SPEED_MINまたはBLOWING_SPEED_MAXの2段階で調節可能です．
        double speed = AromaShooter.BLOWING_SPEED_MIN;
        int ports[] = { 1, 2 };// 1、2番ポートから香りを噴射

        // 指定した条件で香りを噴射します
        as.blow("", durationMilliSec, density, speed, ports);

        /*
         * blowメソッドは噴射が完了するのを待たずにリターンします．
         * また，disconnectメソッドが呼ばれると直ちに噴射が停止するので，噴射が完了するまで一旦スレッドを休止します．
         */
        try {
            Thread.sleep(durationMilliSec);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        // 駆動中の全ての風力源を停止したうえでシリアルポートからアロマシュータを取り出します
        as.disconnect();
	}

	private static void testRS485Device(){
		// アロマシューターが接続中のシリアルポートを指定します（例：COM3）
        String portName = "/dev/ttyUSB0";
        // アロマシューターのプロダクトIDを指定します
        String productId = "ASN1RA0012";
        final int BUFFER_TIME_USB_CONNECTION = 100;

        // アロマシュータークラスのインスタンスを生成します
        AromaShooter as = new AromaShooter(productId, portName);
        try {
            Thread.sleep(BUFFER_TIME_USB_CONNECTION);
        } catch (InterruptedException e1) {
            e1.printStackTrace();
        }

        int durationMilliSec = 3000;// 噴射期間(ミリ秒)
        double density = 0.4;// 香りの濃度(0～1.0)
        // 噴射速度は、BLOWING_SPEED_MINまたはBLOWING_SPEED_MAXの2段階で調節可能です
        double speed = AromaShooter.BLOWING_SPEED_MAX;

        try {
        	int port4[] = { 4 }; 
            as.blow(as.getId(), durationMilliSec, speed, 1, port4);
            // blowメソッド実行後噴射の終了を待たずに制御が戻るため、噴射時間分スレッドを待機させます
            Thread.sleep(durationMilliSec);

            int portToRun[] = {2, 5};
            as.blow(as.getId(), durationMilliSec, speed, portToRun);
            Thread.sleep(durationMilliSec);
            
            int port3[] = { 3 };
            int port6[] = { 6 }; 
            as.blow(as.getId(), durationMilliSec, speed, port3);
            as.blow(as.getId(), durationMilliSec, speed, port6);
            Thread.sleep(durationMilliSec);

            // 濃度制御での噴射も可能です
            int port1[] = { 1 }; 
            as.blow(as.getId(), durationMilliSec, density, speed, port1);
            Thread.sleep(durationMilliSec);

            // ファンを駆動する際はchangeFanStateでファンの駆動/停止を変更してください
            as.changeFanState(AromaShooter.FAN_STATE_ON);
            Thread.sleep(durationMilliSec);
            as.changeFanState(AromaShooter.FAN_STATE_OFF);
        } catch (InterruptedException e) {
            e.printStackTrace();
        } finally {
            as.disconnect();
        }
	}

}


Sample_SDK_v1/src/com/aromajoin/controllersdksample/MainSample.java
/**
 * 
 */
package com.aromajoin.controllersdksample;

import java.util.Iterator;
import java.util.List;
import java.util.Set;

import tinyb.BluetoothDevice;
import tinyb.BluetoothGattService;

import com.aromajoin.aromashooter.*;
import com.aromajoin.aromashooter.AromaShooterBLE.AromaShooterPeripheral;

/**
 * @author Hanh D TRAN
 * © Aromajoin Corporation
 *
 *  This is class provides two test functions for Aroma Shooter USB and RS-485 devices.
 *
 *  Please run to test functions:
 *      - testUSBDevice
 *      - testRS485Device
 */
public class MainSample {

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		// Discover available ports which were connected to Aroma Shooter USB devices
//		Set<String> availableAromaShooterUSBs = AromaShooter.discoverConnectedAromaShooterUSB();
//		Iterator<String> iterator = availableAromaShooterUSBs.iterator();
		// Iterate on valid ports
//		while(iterator.hasNext()){
//			// Perform a test diffuse for each port.
//			String usbPort = iterator.next();
//			testUSBDevice(usbPort);
//		}
		
		// Discover available ports which were connected to Aroma Shooter RS-485 devices
		String[] productIds = {"ASN1RA0011", "ASN1RA0018"};
		Set<String> availableRS485AromaShooters = AromaShooterSerial.discoverConnectedAromaShooterRS485(productIds);
		Iterator<String> iterator = availableRS485AromaShooters.iterator();
		// Iterate on valid ports
		while(iterator.hasNext()){
			// Perform a test diffuse for each port.
			String rs485PortName = iterator.next();
			for(String productId: productIds){
				testRS485Device(productId, rs485PortName);
			}
		}
	}
	
    /**
     * Test to control Aroma Shooter with USB protocol communication
     */
	private static void testUSBDevice(String portName){
		// Specify the serial port to which the aroma shooter is connected (ex: COM3 on Windows or ttyUSB on Mac/Linux).
		// String portName = "/dev/ttyUSB1";

        // Confirm whether the device connected to the specified serial port is the Aroma shooter USB.
        if (AromaShooterSerial.isAromaShooter("", portName)) {
            System.out.println(portName + " The connected device is an aroma shooter．");
        } else {
            System.out.println(portName + " The connected device is not an aroma shooter．");
            return;
        }

        // Create an instance of the Aroma shooter class for USB protocol.
        AromaShooterSerial as = new AromaShooterSerial(portName);
        System.out.println(as.toString());

        // Inject scent process
        // Injection period (ms)
        int durationMilliSec = 1000;

        // Scent concentration (between 0 ~ 1)
        double density = 1;

        // The injection speed can be adjusted in two values: BLOWING_SPEED_MIN or BLOWING_SPEED_MAX.
        double speed = AromaShooterSerial.BLOWING_SPEED_MIN;

        // Array of ports [1,2] want to inject scent that defined by port number in the aroma shooter 
        int ports[] = { 1 };

        // Inject the scent under specified conditions.
        // For USB protocol, productId is not required.
        as.blow("", durationMilliSec, speed, ports);

        /*
         * The blow method returns without waiting for the injection to complete.
         * Also, as injection stops as soon as the disconnect method is called, it pauses the thread until the injection is complete.
         */
        try {
            Thread.sleep(durationMilliSec);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        // Stop all running wind sources and remove the aroma shooter from the serial port.
        as.disconnect();
	}


    /**
     * Test to control Aroma Shooter with RS-485 protocol communication
     */
	private static void testRS485Device(String productId, String portName){
		// Specify the serial port to which the aroma shooter is connected (ex: COM 3).
		// String portName = "/dev/ttyUSB0";

        // Specify the product ID of the aroma shooter which provided in product package.
		//  String productId = "ASN1RA0012";
        final int BUFFER_TIME_RS458_CONNECTION = 100;

        // Create an instance of the Aroma shooter class for USB protocol.
        // Provides ProductId for communicate to RS-485 device.
        AromaShooterSerial as = new AromaShooterSerial(productId, portName);

        // It should be wait for connection time.
        try {
            Thread.sleep(BUFFER_TIME_RS458_CONNECTION);
        } catch (InterruptedException e1) {
            e1.printStackTrace();
        }

        // Inject scent process
        // Injection period (ms)
        int durationMilliSec = 3000;

        // Scent concentration (between 0 ~ 1)
        double density = 0.4;

        // The injection speed can be adjusted in two values: BLOWING_SPEED_MIN or BLOWING_SPEED_MAX.
        double speed = AromaShooterSerial.BLOWING_SPEED_MAX;

        try {
            // Array of ports [4] want to inject scent that defined by port number in the aroma shooter 
        	int port4[] = { 4 }; 
            // Inject the scent under specified conditions.
            // For RS-485 protocol, productId is required.
            as.blow(productId, durationMilliSec, 1, speed, port4);

            // After executing the blow method, since control AS returns without waiting for the end of the injection, 
            // let the thread wait for the injection time.
            Thread.sleep(durationMilliSec);

            int portToRun[] = {2, 5};
            // Blow without aroma density
            as.blow(as.getId(), durationMilliSec, speed, portToRun);
            Thread.sleep(durationMilliSec);
            
            int port3[] = { 3 };
            int port6[] = { 6 };
            // Blow without aroma density
            as.blow(productId, durationMilliSec, speed, port3);
            as.blow(productId, durationMilliSec, speed, port6);
            Thread.sleep(durationMilliSec);

            // Control injection with control concentration (aroma density) is also possible
            int port1[] = { 1 }; 
            as.blow(productId, durationMilliSec, density, speed, port1);
            Thread.sleep(durationMilliSec);

            // When control a fan please change fan fan stop / stop with changeFanState
            as.changeFanState(AromaShooterSerial.FAN_STATE_ON);
			Thread.sleep(durationMilliSec);
            as.changeFanState(AromaShooterSerial.FAN_STATE_OFF);
        } catch (InterruptedException e) {
            e.printStackTrace();
        } finally {
            as.disconnect();
        }
	}
	
	/**
	 * 
	 */
	private static void testBLEDevices() throws InterruptedException {
		System.out.println("Scanning devices....");
    	AromaShooterBLE aromaShooterBLE = AromaShooterBLE.getInstance();
    	List<AromaShooterPeripheral> fileredAromaShooters = aromaShooterBLE.startScanning();
    	System.out.println("Filtered : " + fileredAromaShooters.size());
    	for(AromaShooterPeripheral asp : fileredAromaShooters){
    		System.out.println(asp.toString());
    	}
    	
    	System.out.println("Connecting device....");
    	aromaShooterBLE.connectDevice("78:C5:E5:6D:EA:B1");
    	for(AromaShooterPeripheral connectedASN : aromaShooterBLE.getConnectedDevices()){
    		System.out.println(connectedASN.toString());
    		aromaShooterBLE.blow(aromaShooterBLE.getConnectedDevice(connectedASN.getBleAddress()), 1000, 1, 2);
    		
    		// Get service from UUID
//    		BluetoothDevice aromashooter = aromaShooterBLE.getConnectedDevice(connectedASN.getBleAddress());
//    		BluetoothGattService asnService = aromaShooterBLE.getService(aromashooter.getAddress(), "00001802-0000-1000-8000-00805f9b34fb");
//            if (asnService == null) {
//                System.err.println("This device does not have the service we are looking for.");
//                aromashooter.disconnect();
//                System.exit(-1);
//            }
//          System.out.println("Found service " + asnService.getUUID());
    	}
    	
    	System.out.println("Disconnecting device....");
    	aromaShooterBLE.disconnectDevice("78:C5:E5:6D:EA:B1");
    	
    	for(AromaShooterPeripheral connectedASN : aromaShooterBLE.getConnectedDevices()){
    		System.out.println(connectedASN.toString());
    	}
    	System.out.println("Done.");
	}
}
